README.md
markdown_toc
============

a flexible markdown toc with more custom style && freedom 

## 用法

see demo/lite/index.html

```javascript
<script type="text/javascript" src="../../js/jquery-1.4.4.min.js"></script>
<script type="text/javascript" src="../../src/markdown_toc.js"></script>
<script type="text/javascript" >
$(document).ready(function(){
	$('#tree').markdown_toc({
		is_auto_number:true
	});
});
</script>
```
		
## 依赖

目前只有1个依赖库

- jquery-1.4+

## 定制

### 定制步骤

1. 修改markdown_toc里的compile_headers_with_item函数
1. 按照自己的需求修改返回的html结构，参数item是节点对象

### 定制方法

```javascript
<script type="text/javascript" src="../../js/jquery-1.4.4.min.js"></script>
<script type="text/javascript" src="../../src/markdown_toc.js"></script>
<script type="text/javascript" >
$(document).ready(function(){
	$('#tree').markdown_toc({
		is_auto_number:true,
		compile_headers_with_item:function(item){
			return " <h"+item.level +"><a href='#" +item.id+ "'>" + item.orderd_title + "</a></h"+item.level +">"
		}
	});
});
</script>
```

### item对象

```javascript
item = {
  "id": 2105,
  "level": 1,
  "orderd_title": "21.5. 三部分的关系",
  "origin_title": "三部分的关系",
  "open": true,
  "pId": 21,
  "target": "_self",
  "url": "#2105"
}
```

### item对象属性说明

- id是唯一编号，可以作为锚点id
- level是标题层级
- orderd_title是带有序号的标题
- origin_title是没有序号的原始标题
- open是十分可以展开，如果是，说明它由子节点
- pId是当前节点的父亲节点id
- target是打开方式
- url是锚点的url地址


## Contributing

1. Fork it
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create new Pull Request

## 推荐

- [i5ting_ztree_toc](https://github.com/i5ting/i5ting_ztree_toc) 带有树形结构
- 生成器 [ruby版本tocmd](https://github.com/i5ting/tocmd.gem) 
- 生成器 [node版本i5ting_toc](https://github.com/i5ting/tocmd.npm)
- [gulp task for i5ting-toc](https://github.com/i5ting/gulp-i5ting-toc)


## 版本历史

- v0.1.0 初始化版本

## 欢迎fork和反馈

- write by `i5ting` shiren1118@126.com

如有建议或意见，请在issue提问或邮件

## License

this repo is released under the [MIT
License](http://www.opensource.org/licenses/MIT).

History.md


0.1.1 / 2014-11-26
==================

  * 增加中间件，为基础ztree_toc做准备


0.1.0 / 2014-11-25
==================

  * updated docs/site
  * added readme
  * got everything working
  * initial commit
  
 

src/markdown_toc_lite.js
/**
 * 1.1.1 = 1*100*100 + 1*100 + 1
 * 1.2.2 = 1*100*100 + 2*100 + 3
 *
 * 1 = 0*100 +1
 */
function encode_id_with_array(opts,arr) {
	var result = 0;
  for(var z = 0; z < arr.length; z++ ) {
		result += factor(opts, arr.length - z ,arr[z]);
  }

	return result;
}


/**
 * 1.1.1 = 1*100*100 + 1*100 + 1
 * 1.2.2 = 1*100*100 + 2*100 + 3
 *
 * 1 = 0*100 +1

	1,1 = 100

 */
function get_parent_id_with_array(opts,arr) {
	var result_arr = [];

  for(var z = 0; z < arr.length; z++ ) {
		result_arr.push(arr[z]);
  }

	result_arr.pop();

	var result = 0;
  for(var z = 0; z < result_arr.length; z++ ) {
		result += factor(opts,result_arr.length - z,result_arr[z]);
  }

	return result;
}

function factor(opts ,count,current) {
	if(1 == count) {
		return current;
	}

	var str = '';
	for(var i = count - 1;i > 0; i-- ) {
		str += current * opts.step+'*';
	}

	return eval( str + '1' );
}

;(function($) {
	/*
	 * 根据header创建目录内容
	 */
	function create_toc(opts) {
		$(opts.documment_selector).find(':header').each(function() {
			var level = parseInt(this.nodeName.substring(1), 10);
			var origin_title = $.trim($(this).text());
			
			_rename_header_content(opts,this,level);
			
			_add_header_node(opts,$(this),level,origin_title);
		});//end each
	}

	/*
	 * 渲染header
	 */
	function render_with_headers(opts) {
	 	return compile_headers(opts);
	}
	
	/*
	 * 渲染header
	 */
	function compile_headers(opts) {
		var result = opts._header_nodes;
		var html = '';
		
		for(var i in result){
			var item = result[i];
			// compile with template
			html += opts.compile_headers_with_item(item);
		}
		
	 	return html;
	}


	/*
	 * 将已有header编号，并重命名
	 */
	function _rename_header_content(opts ,header_obj ,level) {
		if(opts._headers.length == level) {
			opts._headers[level - 1]++;
		} else if(opts._headers.length > level) {
			opts._headers = opts._headers.slice(0, level);
			opts._headers[level - 1] ++;
		} else if(opts._headers.length < level) {
			for(var i = 0; i < (level - opts._headers.length); i++) {
			  // console.log('push 1');
			  opts._headers.push(1);
			}
		}

		if(opts.is_auto_number == true) {
			//另存为的文件里会有编号，所以有编号的就不再重新替换
			if($(header_obj).text().indexOf( opts._headers.join('.') ) != -1){

			}else{
				$(header_obj).text(opts._headers.join('.') + '. ' + $(header_obj).text());
			}
		}
	}
	
	/*
	 * create table with head for anchor for example: <h2 id="#Linux基础">Linux基础</h2>
	 * this method can get a headable anchor
	 * add by https://github.com/chanble
	 */
	function _get_anchor_from_head(header_obj){
		var name = header_obj.html();
		var aname = name.split('.');
		var anchor = aname.pop().trim();
		return anchor;
	}

	/*
	 * 给ztree用的header_nodes增加数据
	 */
	function _add_header_node(opts ,header_obj, level, origin_title) {
		var id  = encode_id_with_array(opts,opts._headers);//for ztree
		var pid = get_parent_id_with_array(opts,opts._headers);//for ztree
		var anchor = id;//use_head_anchor.html#第二部分

		// 默认使用标题作为anchor
		if(opts.use_head_anchor == true){
			anchor = _get_anchor_from_head(header_obj);
		}
		
    // 设置锚点id
		$(header_obj).attr('id',anchor);

		log($(header_obj).text());

		opts._header_offsets.push($(header_obj).offset().top - opts.highlight_offset);

		log('h offset ='+( $(header_obj).offset().top - opts.highlight_offset ) );

		opts._header_nodes.push({
			id					: id,
			level				: level,
			pId					: pid ,
			orderd_title				: $(header_obj).text()||'null',
			origin_title: origin_title,
			open				: true,
			url					: '#'+ anchor,
			target			: '_self'
		});
	}

	/*
	 * 根据滚动确定当前位置，并更新ztree
	 */
	function bind_scroll_event_and_update_postion(opts) {
		var timeout;
	    var highlight_on_scroll = function(e) {
			if (timeout) {
				clearTimeout(timeout);
			}

			timeout = setTimeout(function() {
				var top = $(opts.scroll_selector).scrollTop(),highlighted;

				if(opts.debug) console.log('top='+top);

				for (var i = 0, c = opts._header_offsets.length; i < c; i++) {
					// fixed: top+5防止点击ztree的时候，出现向上抖动的情况
					if (opts._header_offsets[i] >= (top + 5) ) {
						console.log('opts._header_offsets['+ i +'] = '+opts._header_offsets[i]);
						$('a').removeClass('curSelectedNode');

						// 由于有root节点，所以i应该从1开始
				  		var obj = $('#tree_' + (i+1) + '_a').addClass('curSelectedNode');
						break;
					}
				}
			}, opts.refresh_scroll_time);
		};

	  if (opts.highlight_on_scroll) {
	    $(opts.scroll_selector).bind('scroll', highlight_on_scroll);
	    highlight_on_scroll();
	  }
	}

	/*
	 * 初始化
	 */
	function init_with_config(opts) {
		opts.highlight_offset = $(opts.documment_selector).offset().top;
	}

	/*
	 * 日志
	 */
	function log(str) {
		if($.fn.ztree_toc.defaults.debug == true) {
			console.log(str);
		}
	}

	$.fn.ztree_toc = function(options) {
		// 将defaults 和 options 参数合并到{}
		var opts = $.extend({},$.fn.ztree_toc.defaults,options);

		return this.each(function() {
			opts._zTree = $(this);

			// 初始化
			init_with_config(opts);

			// 创建table of content，获取元数据_headers
			create_toc(opts);

			// 根据_headers生成ztree
			var r = render_with_headers(opts);


			$(opts._zTree).html("<ul>" + r  +"</ul>");
			// 根据滚动确定当前位置，并更新ztree
		    // bind_scroll_event_and_update_postion(opts);
		});
		// each end
	}

	//定义默认
	$.fn.ztree_toc.defaults = {
		_zTree: null,
		_headers: [],
		_header_offsets: [],
		_header_nodes: [{ id:1, pId:0, orderd_title:"Table of Content",open:true}],
		debug: false,
		/*
		 * 使用标题作为anchor
		 * create table with head for anchor for example: <h2 id="#Linux基础">Linux基础</h2>
		 * 如果标题是唯一的，建议开启此选项，如果标题不唯一，还是使用数字吧
		 * 此选项默认是false，不开启
		 */
		use_head_anchor: false,
    scroll_selector: 'window',
		highlight_offset: 0,
		highlight_on_scroll: true,
		/*
		 * 计算滚动判断当前位置的时间，默认是50毫秒
		 */
		refresh_scroll_time: 50,
		documment_selector: 'body',
		
		/**
		{id: 2105 ,level:1,orderd_title: "21.5. 三部分的关系",origin_title:"三部分的关系", open: true,  pId: 21 ,target: "_self", url: "#2105"}
	
		function compile_headers_with_item(item) {
			return "<li>" + item.name + "<li>"
		}
		**/
		compile_headers_with_item:function(item){
			return " <h"+item.level +"><a href='#" +item.id+ "'>" + item.orderd_title + "</a></h"+item.level +">"
		},
		/*
		 * ztree的位置，默认是在上部
		 */
		is_posion_top: true,
		/*
		 * 默认是否显示header编号
		 */
		is_auto_number: false,
		/*
		 * 默认是否展开全部
		 */
		is_expand_all: true,
		/*
		 * 是否对选中行，显示高亮效果
		 */
		is_highlight_selected_line: true,
		step: 100 
	};

})(jQuery);


src/markdown_toc.js
/**
 * 1.1.1 = 1*100*100 + 1*100 + 1
 * 1.2.2 = 1*100*100 + 2*100 + 3
 *
 * 1 = 0*100 +1
 */
function encode_id_with_array(opts,arr) {
	var result = 0;
  for(var z = 0; z < arr.length; z++ ) {
		result += factor(opts, arr.length - z ,arr[z]);
  }

	return result;
}


/**
 * 1.1.1 = 1*100*100 + 1*100 + 1
 * 1.2.2 = 1*100*100 + 2*100 + 3
 *
 * 1 = 0*100 +1

	1,1 = 100

 */
function get_parent_id_with_array(opts,arr) {
	var result_arr = [];

  for(var z = 0; z < arr.length; z++ ) {
		result_arr.push(arr[z]);
  }

	result_arr.pop();

	var result = 0;
  for(var z = 0; z < result_arr.length; z++ ) {
		result += factor(opts,result_arr.length - z,result_arr[z]);
  }

	return result;
}

function factor(opts ,count,current) {
	if(1 == count) {
		return current;
	}

	var str = '';
	for(var i = count - 1;i > 0; i-- ) {
		str += current * opts.step+'*';
	}

	return eval( str + '1' );
}

;(function($) {
	/*
	 * 根据header创建目录内容
	 */
	function create_toc(opts) {
		$(opts.documment_selector).find(':header').each(function() {
			var level = parseInt(this.nodeName.substring(1), 10);
			var origin_title = $.trim($(this).text());
			
			_rename_header_content(opts,this,level);
			
			_add_header_node(opts,$(this),level,origin_title);
		});//end each
	}

	/*
	 * 渲染header
	 */
	function render_with_headers(opts) {
		opts.render_before(opts);
	 	return compile_headers(opts);
	}
	
	/*
	 * 渲染header
	 */
	function compile_headers(opts) {
		var result = opts._header_nodes;
		var html = '';
		
		for(var i in result){
			var item = result[i];
			// compile with template
			html += opts.compile_headers_with_item(item);
		}
		
		opts.render_after(opts, html);
	 	return html;
	}


	/*
	 * 将已有header编号，并重命名
	 */
	function _rename_header_content(opts ,header_obj ,level) {
		if(opts._headers.length == level) {
			opts._headers[level - 1]++;
		} else if(opts._headers.length > level) {
			opts._headers = opts._headers.slice(0, level);
			opts._headers[level - 1] ++;
		} else if(opts._headers.length < level) {
			for(var i = 0; i < (level - opts._headers.length); i++) {
			  // console.log('push 1');
			  opts._headers.push(1);
			}
		}

		if(opts.is_auto_number == true) {
			//另存为的文件里会有编号，所以有编号的就不再重新替换
			if($(header_obj).text().indexOf( opts._headers.join('.') ) != -1){

			}else{
				$(header_obj).text(opts._headers.join('.') + '. ' + $(header_obj).text());
			}
		}
	}
	
	/*
	 * create table with head for anchor for example: <h2 id="#Linux基础">Linux基础</h2>
	 * this method can get a headable anchor
	 * add by https://github.com/chanble
	 */
	function _get_anchor_from_head(header_obj){
		var name = header_obj.html();
		var aname = name.split('.');
		var anchor = aname.pop().trim();
		return anchor;
	}

	/*
	 * 给ztree用的header_nodes增加数据
	 */
	function _add_header_node(opts ,header_obj, level, origin_title) {
		var id  = encode_id_with_array(opts,opts._headers);//for ztree
		var pid = get_parent_id_with_array(opts,opts._headers);//for ztree
		var anchor = id;//use_head_anchor.html#第二部分

		// 默认使用标题作为anchor
		if(opts.use_head_anchor == true){
			anchor = _get_anchor_from_head(header_obj);
		}
		
    // 设置锚点id
		$(header_obj).attr('id',anchor);

		log($(header_obj).text());

		opts._header_offsets.push($(header_obj).offset().top - opts.highlight_offset);

		log('h offset ='+( $(header_obj).offset().top - opts.highlight_offset ) );

		opts._header_nodes.push({
			id					: id,
			level				: level,
			pId					: pid ,
			orderd_title				: $(header_obj).text()||'null',
			origin_title: origin_title,
			open				: true,
			url					: '#'+ anchor,
			target			: '_self'
		});
	}

	/*
	 * 根据滚动确定当前位置，并更新ztree
	 */
	function bind_scroll_event_and_update_postion(opts) {
		var timeout;
	    var highlight_on_scroll = function(e) {
			if (timeout) {
				clearTimeout(timeout);
			}

			timeout = setTimeout(function() {
				var top = $(opts.scroll_selector).scrollTop(),highlighted;

				if(opts.debug) console.log('top='+top);

				for (var i = 0, c = opts._header_offsets.length; i < c; i++) {
					// fixed: top+5防止点击ztree的时候，出现向上抖动的情况
					if (opts._header_offsets[i] >= (top + 5) ) {
						console.log('opts._header_offsets['+ i +'] = '+opts._header_offsets[i]);
						$('a').removeClass('curSelectedNode');

						// 由于有root节点，所以i应该从1开始
				  		var obj = $('#tree_' + (i+1) + '_a').addClass('curSelectedNode');
						break;
					}
				}
			}, opts.refresh_scroll_time);
		};

	  if (opts.highlight_on_scroll) {
	    $(opts.scroll_selector).bind('scroll', highlight_on_scroll);
	    highlight_on_scroll();
	  }
	}
	/*
	 * 初始化中间件
	 */
	function init_with_middlewares(opts){
		var middlewares = opts.middlewares;
		for(o in middlewares){
			middlewares[o](opts);
		}
	}

	/*
	 * 初始化
	 */
	function init_with_config(opts) {
		opts.highlight_offset = $(opts.documment_selector).offset().top;
	}

	/*
	 * 日志
	 */
	function log(str) {
		if($.fn.markdown_toc.defaults.debug == true) {
			console.log(str);
		}
	}

	$.fn.markdown_toc = function(options) {
		// 将defaults 和 options 参数合并到{}
		var opts = $.extend({},$.fn.markdown_toc.defaults,options);

		return this.each(function() {
			opts._zTree = $(this);

			// 初始化
			init_with_config(opts);
			
			// 初始化middlewares
			init_with_middlewares(opts);

			// 创建table of content，获取元数据_headers
			create_toc(opts);

			// 根据_headers生成ztree
			render_with_headers(opts);

			// 根据滚动确定当前位置，并更新ztree
		    // bind_scroll_event_and_update_postion(opts);
		});
		// each end
	}

	//定义默认
	$.fn.markdown_toc.defaults = {
		_zTree: null,
		_headers: [],
		_header_offsets: [],
		_header_nodes: [{ id:1, pId:0, orderd_title:"Table of Content",open:true}],
		debug: false,
		/*
		 * 使用标题作为anchor
		 * create table with head for anchor for example: <h2 id="#Linux基础">Linux基础</h2>
		 * 如果标题是唯一的，建议开启此选项，如果标题不唯一，还是使用数字吧
		 * 此选项默认是false，不开启
		 */
		use_head_anchor: false,
    scroll_selector: 'window',
		highlight_offset: 0,
		highlight_on_scroll: true,
		/*
		 * 计算滚动判断当前位置的时间，默认是50毫秒
		 */
		refresh_scroll_time: 50,
		documment_selector: 'body',
		render_before:function(opts){
			
		},
		render_after:function(opts,compiled_html){
			$(opts._zTree).html("<ul>" + compiled_html  +"</ul>");
		},
		/*
		 * 中间件
		 */
		middlewares:[
			function(opts){
				console.log('aaaaaa');
			},
			function(opts){
				console.log('bbbbbb');
			}
		],
		/**
		{id: 2105 ,level:1,orderd_title: "21.5. 三部分的关系",origin_title:"三部分的关系", open: true,  pId: 21 ,target: "_self", url: "#2105"}
	
		function compile_headers_with_item(item) {
			return "<li>" + item.name + "<li>"
		}
		**/
		compile_headers_with_item:function(item){
			return " <h"+item.level +"><a href='#" +item.id+ "'>" + item.orderd_title + "</a></h"+item.level +">"
		},
		/*
		 * ztree的位置，默认是在上部
		 */
		is_posion_top: true,
		/*
		 * 默认是否显示header编号
		 */
		is_auto_number: false,
		/*
		 * 默认是否展开全部
		 */
		is_expand_all: true,
		/*
		 * 是否对选中行，显示高亮效果
		 */
		is_highlight_selected_line: true,
		step: 100 
	};

})(jQuery);
